lib/Partitioner/PartitionerUtils.cpp

```diff
@@ -140,72 +139,154 @@ getOutUsersWithOnePredecessor(const std::set<Node *> &nodes) {
   return ret;
 }
 
-uint64_t getOutMemPerNode(const std::set<Node *> &nodes, const Node *node) {
-  uint64_t ret = 0;
-  // Assume that for each Node, the max number of outputs is 5 so far.
-  llvm::SmallSet<size_t, 5> used;
-  for (const auto &U : node->getUsers()) {
-    const Node *user = U.getUser();
-    if (nodes.find(const_cast<Node *>(user)) != nodes.end()) {
-      continue;
+/// \returns the list of users who use \p node's \p idx - th output.
+std::vector<Node *> getNodeUsers(Node *node, size_t idx) {
+  std::vector<Node *> ret;
+  NodeValue nodeVal = node->getNthResult(idx);
+  for (auto &U : node->getUsers()) {
+    Node *user = U.getUser();
+    for (size_t i = 0, e = user->getNumInputs(); i < e; i++) {
+      NodeValue val = user->getNthInput(i);
+      if (val == nodeVal) {
+        ret.push_back(user);
+        break;
+      }
     }
-    // Find the user that doesn't belong to this subgraph.
-    // If a node has several users, the memory usage is only counted once.
-    for (size_t idx = 0, end = user->getNumInputs(); idx < end; idx++) {
-      NodeValue val = user->getNthInput(idx);
-      if (val.getNode() == node && used.count(val.getResNo()) == 0) {
-        ret += node->getType(val.getResNo())->getSizeInBytes();
-        used.insert(val.getResNo());
+  }
+  return ret;
+}
+
+/// \returns the memory usage of the output caused by \p node who has users not
+/// in the set \p nodes.
+uint64_t getOutMemPerNode(const NodesSetTy &nodes, const Node *node) {
+  uint64_t ret = 0;
+  for (size_t i = 0, e = node->getNumResults(); i < e; i++) {
+    std::vector<Node *> users = getNodeUsers(const_cast<Node *>(node), i);
+    for (auto &user : users) {
+      if (nodes.find(const_cast<Node *>(user)) == nodes.end()) {
+        ret += node->getType(i)->getSizeInBytes();
```

Comment:
nit: If you create a `NodeValue NV(node, i)` to pass above to `getNodeUsers()`, then you can also just do `NV.getType()` here.

Reply:
Now I just passed` node->getNthResult(i)` to getNodeUsers, so I think it is OK to use `node->getType(i)` without introducing another var:)